dask/local.py

```diff
@@ -487,7 +489,7 @@ def fire_tasks(chunksize):
                             pack_exception,
                         )
                     )
-                del state["ready"][:ntasks]
+                del state["ready"][nready - ntasks : nready]
```

Comment:
Maybe this would be more natural as the following?

```python
while len(state["ready"]) > ...:
    key = state["ready"].pop()
```

Reply:
This goes along with the assumption I mention above. Does `state` get updated by other threads/asynchronous routines? If so then doing `pop` could be removing things that were being added by these other methods. That's why I did the fancy `nready`-based indexing. Otherwise, if that assumption is wrong then the original `[:ntasks]` would work.

Regarding `while` loop, that would perform worse compared to `del` wouldn't it?

Reply:
>  Otherwise, if that assumption is wrong then the original [:ntasks] would work.

Sorry, no it wouldn't it would have to be `del state["ready"][-ntasks:]`.

Reply:
> Does state get updated by other threads/asynchronous routines?

The scheduler code is all run from within the same thread.  While Dask enables multi-threaded execution it has historically been hard to get any speedup in data-structure heavy code like this.

> Regarding while loop, that would perform worse compared to del wouldn't it?

I doubt that it would perform worse in any measurable way.  I think that it's also a more common pattern.  

I think that what you're trying to do is to pull tasks off the end of the ready stack until we're saturated again.  I think that `while True: L.pop()` is a more common idiom there.  

When I see `ready[nready - ntasks:nready]` my brain has to think a lot more to figure out what's going on.

Reply:
> When I see ready[nready - ntasks:nready] my brain has to think a lot more to figure out what's going on.

How does `del state["ready"][-ntasks:]` make you feel?

I'll see how I can clean this up...maybe time the while loop approach. I personally (Computer Engineering background) would see the `while` loop and wonder why `del` wasn't used, but not my project so :shrug: 

This also ignores the failing tests that don't like that I'm processing things in reverse order. I don't have the historical context to know why that test exists and why execution order matters.

Reply:
> I'll see how I can clean this up...maybe time the while loop approach. I personally (Computer Engineering background) would see the while loop and wonder why del wasn't used, but not my project so 🤷

`while condition: L.pop()` is pretty common in Dask code.  I recommend that we use that idiom for local consistency.

> I don't have the historical context to know why that test exists and why execution order matters.

As you've discovered, execution order matters deeply for memory usage :) I recommend checking out git blame for the test and seeing if anything comes up there.  Maybe it was added / modified the last time someone edited this code.  Regardless, we should add tests to serve as a strong signal that execution order should follow dask.order ordering.  Maybe a Callback that ensures that the order/priority of executed tasks in a single-threaded situation match dask.order?

Reply:
Aw, I just finished timing it:

```python
In [10]: %%timeit
    ...: state = {"ready": list(range(100))}
    ...: ntasks = 10
    ...: nready = len(state["ready"])
    ...: while len(state["ready"]) > nready - ntasks: state["ready"].pop()
    ...:
    ...:
1.94 µs ± 46.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [11]: %%timeit
    ...: state = {"ready": list(range(100))}
    ...: ntasks = 10
    ...: nready = len(state["ready"])
    ...: del state["ready"][-ntasks:]
    ...:
    ...:
685 ns ± 4.22 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

I don't use timeit a lot so I put all the setup code in the timed code to make sure that the list didn't go empty and change the results of the test.

Reply:
> I personally (Computer Engineering background) 

I just realized how arrogant this sounds. Not how I meant it. Just saying it may explain a differing opinion on the matter.

Reply:
> Maybe it was added / modified the last time someone edited this code. Regardless, we should add tests to serve as a strong signal that execution order should follow dask.order ordering. Maybe a Callback that ensures that the order/priority of executed tasks in a single-threaded situation match dask.order?

Last change to the `test_local.py` ordering test was you in #3271. So that means `dask.order` doesn't work well for this example case (220MB versus 10GB memory usage). I see priorities can be controlled (https://distributed.dask.org/en/latest/priority.html), but is there any other way to order things so reductions like this calculation work better?

```python
d = da.random.random((5000, 250000), chunks=(1, -1)).sum()
```

Let me try turning graph optimization back on...

Reply:
Yeah, 1us of difference doesn't matter much here.  Getting an item off of a Queue takes 50us.